server/grammar/Yml.g4

```diff
@@ -1,458 +1,482 @@
+;
+argsBlock: ARGS OPEN_BRACE variableBlockContent CLOSE_BRACE;
+localBlock: LOCAL OPEN_BRACE variableBlockContent CLOSE_BRACE;
+staticBlock: STATIC OPEN_BRACE staticDeclaration* CLOSE_BRACE;
+
+methodDeclaration: methodIntro memberOption=field*;
+
+/*
+    // accepts structures like:
+	method aggregateTree
+		args {
+			Symbol 				aggMode
+			--> key 			_AGGREGATION_MODE
+			--> implicitKey 	true
+			--> defaultValue 	AGGREGATION_SUM
+	}
+	--> domains Void
+*/
+methodIntro: (FUNCTION | METHOD | TEXT_METHOD | TEXT_FUNCTION) ymlId
+    (
+        argsBlock
+        | OPEN_PAR argumentList CLOSE_PAR
+    )
+;
+
+argumentList: mandatoryArgs? optionalArgs?;
+
+mandatoryArgs: mandatoryArgDecl (COMMA mandatoryArgDecl)*;
+
+mandatoryArgDecl: argType=ymlId argName=ymlId argOptionList?;
+
+optionalArgs:
+    COMMA? OPEN_BRACE
+    (
+        optionalArguments+=optionalArgDecl
+        (
+            COMMA optionalArguments+=optionalArgDecl
+        )*
+    )? CLOSE_BRACE argSetName=ymlId?
+;
+
+optionalArgDecl:
+    (
+        OPEN_BRACKET optionalKeyName=ymlId CLOSE_BRACKET
+        | mandatoryKeyName=ymlId
+    ) COLON argType=ymlId argName=ymlId? argOptionList?
+;
+
+argOptionList:
+    OPEN_BRACE arguments+=argOptionBlock (COMMA arguments+=argOptionBlock)* CLOSE_BRACE
+;
+
+argOptionBlock: argOption=ymlId optionValue=value?;
+
+order1Block: instanciationCondition conditionBlock?;
+
+instanciationCondition: inlineOperation;
+
+order1FullCondition: conditionBlock? order1Block*;
+
+//Comparisons
+combinedComparison:
+    OPEN_PAR combinedComparison CLOSE_PAR
+    | leftCondition=combinedComparison COND_AND rightCondition=combinedComparison
+    | leftCondition=combinedComparison COND_OR rightCondition=combinedComparison
+    | comparison
+;
+
+comparison: leftValue=value comparisonOperator rightValue=value;
+
+comparisonOperator:
+    EQUAL_COMP
+    | NOT_EQUALS
+    | LESS_OR_EQUAL
+    | GREATE_OR_EQUAL
+    | STRICT_LESS
+    | STRICT_GREAT
+;
+
+instruction_multivaluedAssignment:
+    leftHand=value MULTIVALUED_ASSIGNMENT rightHand=value
+;
+instruction_assignment: leftHand=value EQUAL_ASSIGNMENT rightHand=value;
+
+conditionBlock: order0Condition+;
+
+order0Condition: combinedComparison | existentialOperator;
+
+instruction_switchCase_withValue:
+    SWITCH OPEN_PAR value CLOSE_PAR OPEN_BRACE instructionCase* instructionDefault? CLOSE_BRACE
+;
+instruction_switchCase_asIf:
+    SWITCH OPEN_BRACE instructionCase* instructionDefault? CLOSE_BRACE
+;
+instructionCase:
+    CASE OPEN_PAR valueOrCondition CLOSE_PAR COLON (instruction | actionBlock)
+;
+instructionDefault: DEFAULT COLON (instruction | actionBlock);
+instruction_break: BREAK SEMICOLON?;
+
+instruction_ifElse: instruction_if (ELSE (actionBlock | instruction))?;
+instruction_if:
+    IF OPEN_PAR order0Condition CLOSE_PAR (actionBlock | instruction)
+;
+instruction_forall:
+    FORALL OPEN_PAR (ymlId | instanciationVariable) IN (value | FUNCTION) CLOSE_PAR
```

Comment:
I was expecting `functionCall` rather than `FUNCTION` :thinking: So you can write something like…

```
forall (foo in function)
```

?

:thinking:

Reply:
No, in this case, we do accept the `FUNCTION` token too. This is clearly intended.

Reply:
Then I'd like to see an YML code example, because this is puzzling xD

Reply:
@mminot-yseop I explained what it does.

Reply:
I'm starting to understand, but why are functions the _only_ type that needs to be explicitly mentioned in this rule? :thinking: What about `Class`, for example, or stuff like this?

Reply:
We don't have a token explicitly defined for `Class`, while we have it for `Function`. Therefore the type of the token receiving the text `Function` will be of type `FUNCTION` and not of type `YMLID` as expected in the rule `value`.

Reply:
So `Class` will count as a “value” and match that part of the rule, is that it?

Reply:
And the use of `FUNCTION` just adds information to the syntactic tree that get built?

Reply:
> So Class will count as a “value” and match that part of the rule, is that it?

Yes, exactly.

> And the use of FUNCTION just adds information to the syntactic tree that get built?

If the parser doesn't have this information, the parsing will fail.